### ISSUE TITLE
twistd --interactive (or something) that starts up an async REPL

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11334]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#789|
|Type|enhancement|
|Created|2004-11-21 11:01:12Z|

Attachments:

* [manhole-test.log](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fc1/fc1200c7a7aa52109d762a9f005b149abef01479/f63b8d3e3d3ee469f4f6024ed56a329b08b57ff3.log) (116279 bytes) - added by exarkun on 2004-11-22 06:12:12Z - 
* [manhole-trial.out](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fc1/fc1200c7a7aa52109d762a9f005b149abef01479/db680be2555ec4f18a4aec39f60c67354692088b.out) (45099 bytes) - added by exarkun on 2004-11-22 06:12:38Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__789 789
type__enhancement enhancement
reporter__radix radix
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__ 
component__core core
keywords__twistd twistd
time__1101034872000000 1101034872000000
changetime__1315571409000000 1315571409000000
version__ 
owner__ 
cc__radix cc__exarkun cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
This is actually already implemented in exarkun's sandbox, but I really really
really would love exarkun a whooooooooole lot if he merged it into twisted core
before 2.0.

The current blocking issue was how to get it into twisted while remaining
backwards compatible with existing manhole or other stuff. My opinion at this
point is "whatever". I don't personally care about backwards compatibility so I
wouldn't mind if the old stuff was thrown out, and I also wouldn't mind if this
was just put into a new package/module. I think the feature is too important to
be blocked by that stuff. :-)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
The plan is to move this code into twisted.conch.

The real hold up is a number of failing tests.
Relevant output attached.

Reproduce by running `trial -v test_*.py' in Twisted/sandbox/exarkun/manhole/
against current SVN trunk@HEAD.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Created /branches/exarkun/manhole/ and moved the code out of the sandbox and
into twisted/conch/.  There are still some problems:

  the module layout is probably wrong, things are kind of scattered around.  

  some ssh tests are still failing

  there is a race condition in the stdio tests

  insert/typeover mode has a bit of confusion in it that needs to be cleared up

Other than this, things are looking pretty good.  I think someone can start
writing the twistd code to handle --iteractive now.  I'd do it, but twistd is a
scary place that I don't particularly want to be in right now.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I had another look at this recently, and was going to give it a try, but it
appears the example for doing manhole on STDIO has disappeared. I only see the
SSH and telnet ones. Does it exist any more?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
twisted/conch/stdio.py
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Thanks. :-)

Given that the proposed feature would make core depend on conch, I decided to
write an alternate script entirely that wraps twistd with this feature. Might as
well put that ability to subclass ServerOptions to good use.

It's implemented in branches/radix/ctwistd. I may make it also change the
default log observer so that, when the logfile is stdout, it puts a newline
before each message, or something.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
You'll need to do more than put a newline before each message.  If a message is
logged while someone is midway through typing a line, you'll have to restore the
state of the line on the display after the log message thrashes it.

I think this should be a *twistd* feature, provided conditionally based on the
presence of conch.  I think logging should be handled using a pluggable log
observer scheme, support for which should be added to twistd.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I initially wanted it to be a feature of twistd that depended on conch, but that
made me feel slightly uneasy, and the implementation was still slightly hard
after about 20 minutes of hacking about (jesus christ twistd and t.a.app are a
mess). I think ctwistd is a pretty elegant solution, especially given it's so small.

Is it really that important for it to be a feature of twistd? if you think so, I
can reimplement it. btw, open up another bug for a pluggable log observer
thingy, if you want.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
I'd rather it be a feature of twistd, but having ctwistd is better than what was
there before.

We really need to fix twistd.  Soon.  Things like ctwistd and axiomatic are
severely hampered by the current architecture (most obviously: axiomatic gains
no benefits from ctwistd).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

This would be a really great feature to lead with in a release announcement.  I totally want it.  It depends on [#2991](https://github.com/twisted/twisted/issues/2991).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Maybe this should be a separate ticket, but here's another idea that would be nice.

My main use-case for this, of course, is development Axiom instances where I want to poke around at what's going on.  Of course this would be most useful if the axiom plugin could populate the interactive interpreter with convenient short aliases for things I will commonly want to access.  It would be nice to have an additional plug-in method which could allow twistd plugins to populate the interactive namespace.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

I can't see any reason this needs to be targeted at 9.0 (8.2+1).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_15"></a>|@MostAwesomeDude commented|
|-|-|

Does "python -mtwisted.conch.stdio" cover this? If not, would it be possible to connect that functionality to twistd somewhat?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Just some random thoughts...  It's hard to do this as a twistd subcommand, because subcommands don't have good control over logging, and a repl needs to be able to fiddle with logging to keep stdio clean.  Also, twistd is in core and twisted.conch.stdio is (clearly) in Conch, so we need to be careful about dependencies.  Ideally twistd wouldn't know about anything from Conch (which is why we have subcommand plugins).
